File: cpp/55. jump-game.cpp
//
// Created by Rutvik Patel on 1/30/24.
//

#include "include.h"

// solution

class Solution {
public:
    bool canJump(std::vector<int>& nums) {
        int idx = 0, max_pos = 0;

        while (idx <= max_pos) {
            max_pos = std::max(max_pos, idx + nums[idx]);
            if (max_pos >= nums.size() - 1) return true;
            ++idx;
        }

        return false;
    }
};

// main

int main(__attribute__((unused)) int argc, __attribute__((unused)) char *argv[]) {
    std::vector<std::pair<std::vector<int>, bool>> test_cases = {
        { { 2, 3, 1, 1, 4 }, true },
        { { 3, 2, 1, 0, 4 }, false },
        { { 0 }, true },
    };

    Solution solution;

    for (auto& [args, ans] : test_cases)
        assert_m(solution.canJump(args) == ans, "Failed!");

    log("All testcases passed!");
}

File: python/55. jump-game.py
from typing import List


# lc sol start
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        distance = len(nums)
        maxPos = 0
        i = 0
        while i <= maxPos:
            maxPos = max(maxPos, i + nums[i])
            if maxPos >= distance - 1:
                return True
            i += 1

        return False


# lc sol end


def main() -> None:
    test_cases = {
        'inputs': [([2, 3, 1, 1, 4],), ([3, 2, 1, 0, 4],)],
        'outputs': [True, False],
    }

    solution = Solution()

    for idx, args in enumerate(test_cases['inputs']):
        op = solution.canJump(*args)
        assert op == test_cases['outputs'][idx], \
            f"expected output {test_cases['outputs'][idx]} for input {args}, got {op}"

    print('all sample test cases passed successfully!')


if __name__ == '__main__':
    main()
